File: Ej1/src/practica1ds/Carrera.java
package practica1ds;

import java.util.ArrayList;

public abstract class Carrera implements Runnable {
    protected String tipo;
    protected ArrayList<Bicicleta> bicicletas;

    public Carrera(String tipo) {
        this.tipo = tipo;
        this.bicicletas = new ArrayList<Bicicleta>();
    }

    public Carrera(ArrayList<Bicicleta> bicicletas) {
        this.bicicletas = bicicletas;
    }

    @Override
    public void run() {
    }

    public void retirarBicicletaAleatoria() {
        int aleatorio = (int) (Math.random() * bicicletas.size());
        bicicletas.remove(aleatorio);
        System.out.println("Se ha retirado la bicicleta nº: " + aleatorio + " de la carrera de " + tipo + "\n");
    }

    public void agregarBicicleta(Bicicleta bicicleta) {
        bicicletas.add(bicicleta);
    }
}

File: Ej2/src/carrera.py
from abc import ABC, abstractmethod
import random


class Carrera(ABC):

    def __init__(self, tipo, bicicletas=[]):
        self.tipo = tipo
        self.bicicletas = bicicletas

    def agregar_bicicleta(self, bicicleta):
        if bicicleta.get_tipo == self.__tipo:
            self.bicicletas.append(bicicleta)
            return True
        else:
            print (f'La bicicleta {bicicleta.identificador} no es de tipo {self.__tipo}')
            return False

    def __retirar_bicicleta_aleatoria(self):
        bicicleta = random.choice(self.bicicletas)
        self.bicicletas.remove(bicicleta)
        return bicicleta

    @abstractmethod
    def iniciar_carrera(self):
        pass